src/audio/igo_nr/igo_nr.c

```diff
@@ -69,179 +69,316 @@ static void igo_nr_lib_process(struct comp_data *cd)
 	}
 }
 
+static void *igo_cir_wrap(void *ptr, void *cir_buf_end, size_t cir_buf_size)
+{
+	if (ptr > cir_buf_end)
+		ptr = (char *)ptr - cir_buf_size;
+
+	return ptr;
+}
+
 #if CONFIG_FORMAT_S16LE
-static void igo_nr_capture_s16(struct comp_data *cd,
-			       const struct audio_stream *source,
-			       struct audio_stream *sink,
-			       int32_t frames)
+static int igo_nr_capture_s16(struct comp_data *cd,
+			      struct sof_source *source,
+			      struct sof_sink *sink,
+			      int32_t frames)
 {
-	int32_t nch = audio_stream_get_channels(source);
-	int32_t i;
-	int32_t j;
-	int32_t idx_in = 0;
-	int32_t idx_out = 0;
-	int16_t *x;
-	int16_t *y;
+	int16_t *x, *y1, *y2;
+	int16_t *x_start, *y_start, *x_end, *y_end;
+	int16_t sample;
+	size_t x_size, y_size;
+	size_t request_size = frames * source_get_frame_bytes(source);
+	int sink_samples_without_wrap;
+	int samples_without_wrap;
+	int samples_left;
+	int ret;
+	int nch = source_get_channels(source);
+	int i, j, k;
+
 #if CONFIG_DEBUG
 	int32_t dbg_en = cd->config.igo_params.dump_data == 1 && nch > 1;
 	int32_t dbg_ch_idx;
+
+	/* Under DEBUG mode, overwrite the next channel with input interleavedly. */
+	if (cd->config.active_channel_idx >= nch)
+		dbg_ch_idx = 0;
+	else
+		dbg_ch_idx = cd->config.active_channel_idx + 1;
+
 #endif
 
+	ret = source_get_data(source, request_size, (void const **)&x,
+			      (void const **)&x_start, &x_size);
+	if (ret)
+		return ret;
+
+	ret = sink_get_buffer(sink, request_size, (void **)&y1, (void **)&y_start, &y_size);
+	if (ret) {
+		source_release_data(source, 0);
+		return ret;
+	}
+
+	y2 = y1;
+	x_end = x_start + x_size;
+	y_end = y_start + y_size;
+
 	/* Deinterleave the source buffer and keeps the active channel data as input. */
-	for (i = 0; i < frames; i++) {
-
-		/* Pass through all the other channels */
-		for (j = 0; j < nch; j++) {
-			if (j == cd->config.active_channel_idx)
-				continue;
-			x = audio_stream_read_frag_s16(source, idx_in + j);
-			y = audio_stream_write_frag_s16(sink, idx_in + j);
-			*y = (*x);
+	i = 0;
+	samples_left = nch * frames;
+	while (samples_left) {
+		samples_without_wrap = x_end - x;
+		samples_without_wrap = MIN(samples_left, samples_without_wrap);
+		sink_samples_without_wrap = y_end - y1;
+		samples_without_wrap = MIN(samples_without_wrap, sink_samples_without_wrap);
+		for (j = 0; j < samples_without_wrap; j += nch) {
+			for (k = 0; k < nch; k++) {
+				sample = *x;
+				if (k == cd->config.active_channel_idx)
+					cd->in[i] = sample;
+				else
+					*y1 = sample;
+
+				x++;
+				y1++;
+			}
+			i++;
 		}
 
-		x = audio_stream_read_frag_s16(source, idx_in + cd->config.active_channel_idx);
-		cd->in[i] = (*x);
-
-		idx_in += nch;
+		x = igo_cir_wrap(x, x_end, x_size);
```

Comment:
based on Marcin's pipe 2.0, at least, input is linear, output maybe not(if it is the last module).

Reply:
I need this with my tests with TGL with LL pipelines. Let's remove circular capability when we know it's safe to do and the condition you mentioned is acceptable. The overhead from this is wrap minimal.